### ISSUE TITLE
MaxPool2d Output Size Issue

### ISSUE THREAD

**Issue by RoyiAvital** [pytorch/pytorch#6842]
## System Information
- PyTorch 0.3.1
- OS: Linux Mint 18.3
- PyTorch version: 0.3.1
- How you installed PyTorch (conda, pip, source): Conda
- Python version: 3.6 (Anaconda 5.1)
- CUDA/cuDNN version: CUDA 8.0 / CuDNN 7.0.5
- GPU models and configuration: nVidia GTX 1060

## Code
This code reproduce the issue:

```
maxPoolLayer = nn.MaxPool2d(2)
mA = Variable(torch.randn(1, 1, 25, 25))
mA.size()
mB = maxPoolLayer(mA)
mB.size()
```
The code output:
```
torch.Size([1, 1, 12, 12])
```

## Description

According to the [`MaxPool2d()` documentation](http://pytorch.org/docs/master/nn.html#torch.nn.MaxPool2d) if the size is 25x25 and kernel size is 2 the output should be 13 yet as seen above it is 12 ( `floor( ((25 - 1) / 2) + 1 ) = 13` ).  
It seems the last column / row is totally ignored (As input is 24 x 24).

Moreover, the example in documentation won't work as it is missing conversion from `torch.float` to variable.

**Comment by SsnL** [comment]
1. The result is correct because you are missing the dilation term.
2. You are looking at the doc for PyTorch master. Notice the topleft logo says "UNSTABLE". On master the doc example works.

**Comment by fmassa** [comment]
Also, if you want the last column to be used instead of ignored, pass the `ceil_mode=True` option to max pool

**Comment by RoyiAvital** [comment]
@SsnL ,
The [MaxPool2d Documentation of 0.3.1](http://pytorch.org/docs/0.3.0/nn.html#torch.nn.MaxPool2d) has the same formula (Just displayed differently).

Let's go through parameters (Horizontal):

Hin = 25.
Padding = 0.
Dilation = 1.
KernelSize = 2.
Stride = 1.

So we have:

`floor( ((Hin + 2 * Padding - Dilation * (KernelSize -1)) / Stride) + 1 ) = floor( ((25 + 2 * 0 - 1 * (2 - 1)) / 2) + 1 ) = floor( (24 / 2) + 1 ) = 13`.

So either the documentation is wrong or the the code isn't doing as written.
It doesn't make sense to have output of size 12 for input of 25 as it means a complete column is ignored.

Indeed the example in the 0.3.x branch include the Variable function as needed

**Comment by SsnL** [comment]
@RoyiAvital Then you are missing a `-1` after the dilation term.

**Comment by RoyiAvital** [comment]
@SsnL,
Indeed.
Stand corrected.

So it really just ignores part of the input and work on the left / top even part of it?
What exactly happens to the last column in 25 columns image?

**Comment by fmassa** [comment]
It is expected that the last column is dropped. The same happens with convolution. But we have added a parameter to max pooling and average pooling to take the last column into account as I mentioned before.
Looking at the implementation in https://github.com/pytorch/pytorch/blob/master/aten/src/THNN/generic/SpatialDilatedMaxPooling.c#L50 , there seems to be am extra -1 that is not in your equation. I haven't checked the doc to compare though

**Comment by fmassa** [comment]
Also, because you missed the -1, the rounding does make a difference

**Comment by SsnL** [comment]
The last column and row are ignored. You can think of this as a special case of 'valid' convlolution. You can pad the input using the padding layers, or using the `padding=` argument to add the padding. However, notice that the `padding=` only supports symmetrical padding. So if you do `padding=1`, you actually will have the first column (and row) max pooled with zeros, and the last column (and row) max pooled with the second last.
